nixos/attestation-server/test.nix

```diff
@@ -12,18 +12,19 @@ import "${pkgs.path}/nixos/tests/make-test-python.nix" ({ pkgs, ... }: {
       device = "crosshatch";
       signatureFingerprint = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA";
       avbFingerprint = "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB";
-      # TODO: Uncomment when https://github.com/danielfullmer/robotnix/issues/80 is resolved
-      # email = {
-      #   host = "example.com";
-      #   username = "test";
-      #   passwordFile = "${pkgs.writeText "fake-password" "testing123"}"; # NOTE: Don't use writeText like this with a real password!
-      # };
+      email = {
+        host = "example.com";
+        username = "test";
+        passwordFile = "${pkgs.writeText "fake-password" "testing123"}"; # NOTE: Don't use writeText like this with a real password!
+      };
       nginx.enable = false;
     };
   };
 
   testScript = ''
     machine.wait_for_unit("attestation-server.service")
     machine.wait_until_succeeds("curl http://127.0.0.1:8085/")
+    # For some reason the attestation server returns 404. If this could be
+    # fixed we could use curl --fail.
```

Comment:
IIRC, the 404 is correct behavior here.  The interesting endpoints of this service are under `/api/`.  The static HTML is served by nginx, not by the java AttestationServer program.  So, I was just using this to check that it's listening and responding to HTTP requests.

Reply:
Right, I removed the comment.  I also thought about submitting a test attestation request but did not yet dig through the code to find out how.

Reply:
One of the things I intend to do soon-ish is to get Android tests (including some of the Compatibility Test Suite) working inside Nix using the cuttlefish virtualized android device, so we could run android tests in a similar way we do NixOS tests.  If we're *very* lucky we could even do combined end-to-end tests of Android + NixOS apps/services, potentially even Auditor/AttestationServer. This is speculative, and it could be a lot of engineering effort, so don't let that stop you if you want to develop a more isolated test of some of the AttestationServer.

Reply:
I found a way to at least get a challenge from the attestation server:
```patch
--- a/nixos/attestation-server/test.nix
+++ b/nixos/attestation-server/test.nix
@@ -24,5 +24,6 @@ import "${pkgs.path}/nixos/tests/make-test-python.nix" ({ pkgs, ... }: {
   testScript = ''
     machine.wait_for_unit("attestation-server.service")
     machine.wait_until_succeeds("curl http://127.0.0.1:8085/")
+    machine.succeed("curl -fsSL -X POST http://127.0.0.1:8085/challenge | hexdump -C")
   '';
 })
```
I've also had a look at verification but that is highly non-trivial, because we'd have to non-interactively register a user account, extract the fingerprint that is normally displayed as a QR code, generate a keypair, sign the challenge and then verify it.
https://github.com/GrapheneOS/Auditor/blob/b0840cc6cf55230b1a943f855088dbeb3c9a7a4e/app/src/main/java/app/attestation/auditor/RemoteVerifyJob.java#L129-L181